<a href="https://colab.research.google.com/github/jadisiii/DS-Unit-1-Sprint-2-Statistics/blob/master/Ray_Arbizu_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [17]:
### YOUR WORK HERE
import pandas as pd
import numpy as np


#'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
#'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.names'



In [18]:
cols = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

In [19]:
car_imports = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data',
                          header=0,
                          names=cols)

In [20]:
car_imports.sample(3)

symboling normalized-losses   make  ... city-mpg highway-mpg  price
37           0               106  honda  ...       27          33   9095
203         -1                95  volvo  ...       19          25  22625
52           1               113  mazda  ...       31          38   6695

[3 rows x 26 columns]

In [21]:
car_imports.isnull().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [22]:
from scipy.stats import ttest_ind

In [23]:
### YOUR WORK HERE
sample1 = car_imports.sample(30).head(10)

In [24]:
sample1


symboling normalized-losses           make  ... city-mpg highway-mpg  price
164          1               168         toyota  ...       26          29   9298
157          0                91         toyota  ...       34          36   7898
203         -1                95          volvo  ...       19          25  22625
59           0               115          mazda  ...       26          32   8495
138          2                83         subaru  ...       26          31   7053
137          2                83         subaru  ...       31          36   5118
190          0                 ?     volkswagen  ...       19          24  13295
108          0                 ?         peugot  ...       19          24  12440
69          -1                93  mercedes-benz  ...       22          25  31600
80           3               153     mitsubishi  ...       25          32   8499

[10 rows x 26 columns]

In [25]:
print(np.mean(sample1['highway-mpg']))

29.4


In [26]:
results_hmpg = ttest_ind(car_imports['highway-mpg'], sample1['highway-mpg'])
results_hmpg

Ttest_indResult(statistic=0.620042297841059, pvalue=0.5358958302521523)

In [27]:
ttest_ind(car_imports['highway-mpg'], sample1['highway-mpg']).pvalue

0.5358958302521523

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [28]:
### YOUR WORK HERE
print(np.mean(sample1['highway-mpg']))

29.4


Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [29]:
### YOUR WORK HERE
sample2 = car_imports.sample(30).head(100)
print(np.mean(sample2['highway-mpg']))

31.4


Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [44]:
### YOUR WORK HERE

vote_cols = ['party', 'handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']

vote_data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data',
                        header=0,
                        names=vote_cols,
                        na_values = '?')

In [45]:
vote_data.head()

party  ... export-administration-act-south-africa
0  republican  ...                                    NaN
1    democrat  ...                                      n
2    democrat  ...                                      y
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [46]:
vote_data = vote_data.replace({'y':1, 'n':0})

In [47]:
vote_data.head()

party  ...  export-administration-act-south-africa
0  republican  ...                                     NaN
1    democrat  ...                                     0.0
2    democrat  ...                                     1.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [48]:
vote_data['party'].value_counts()

democrat      267
republican    167
Name: party, dtype: int64

In [49]:
rep = vote_data[vote_data['party']=='republican']
rep.shape

(167, 17)

In [50]:
dem = vote_data[vote_data['party']=='democrat']
dem.shape

(267, 17)

In [51]:
print(rep['handicapped-infants'].mean())
print(dem['handicapped-infants'].mean())

0.18902439024390244
0.6046511627906976


### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [61]:
# import
from scipy.stats import ttest_1samp

In [76]:
print(rep['crime'].mean())
print(dem['crime'].mean())

0.98125
0.35019455252918286


In [75]:
### YOUR WORK HERE
# Null Hypothosis: There is no difference between average voting rates (levels of support) for the crime bill between democrats and republicans in the house of representatives. (support is equal)
# Alterntaive Hypothesis: The levels of support between the two parties differ
# 95% Confidence Level

ttest_ind(rep['crime'], dem['crime'], nan_policy='omit')

Ttest_indResult(statistic=16.288201256755894, pvalue=1.796481827173887e-46)

With a pvalue of 1.796481827173887e-46, being < 0.01 I reject
 the null hypothosis that there is no difference between average voting rate form the crime bill between dem's and rep's

With a 99% confidence level I support the alternative hypothosis that levels of support differ between the two parties. 

In [77]:
ttest_ind(dem['export-administration-act-south-africa'], rep['export-administration-act-south-africa'], nan_policy='omit')

Ttest_indResult(statistic=6.893205322931943, pvalue=2.8060959129817107e-11)

With a pvalue of 2.8060959129817107e-11, being < 0.01 I reject the null hypothosis that there is no difference between average voting rate form the export-administration-act-south-africa bill between dem's and rep's

With a 99% confidence level I support the alternative hypothosis that levels of support differ between the two parties.

In [78]:
ttest_ind(rep['water-project-cost-sharing'], dem['water-project-cost-sharing'], nan_policy='omit')

Ttest_indResult(statistic=0.02491808700047811, pvalue=0.9801332440121653)

With a pvalue of 0.9801332440121653, being > 0.01 I fail to reject the null hypothosis that there is no difference between average voting rate form the water-project-cost-sharing bill between dem's and rep's

There seems to be little difference between the dem' and rep's voting on this issue

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [88]:
### YOUR WORK HERE
vote_cols = ['handicapped-infants', 'water-project-cost-sharing', 'adoption-of-the-budget-resolution', 'physician-fee-freeze', 'el-salvador-aid', 'religious-groups-in-schools', 'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile', 'immigration', 'synfuels-corporation-cutback', 'education-spending', 'superfund-right-to-sue', 'crime', 'duty-free-exports', 'export-administration-act-south-africa']
for x in vote_cols:
  print(ttest_ind(rep[x], dem[x], nan_policy='omit'))




Ttest_indResult(statistic=-9.15392122841775, pvalue=2.4195550274149624e-18)
Ttest_indResult(statistic=0.02491808700047811, pvalue=0.9801332440121653)
Ttest_indResult(statistic=-23.12119107755175, pvalue=6.013425749068062e-77)
Ttest_indResult(statistic=49.213324889131904, pvalue=1.0401154557088818e-176)
Ttest_indResult(statistic=21.06444879561732, pvalue=1.3035583587179729e-67)
Ttest_indResult(statistic=9.694471882817886, pvalue=3.415425248481181e-20)
Ttest_indResult(statistic=-12.457095175493478, pvalue=1.6414204749130307e-30)
Ttest_indResult(statistic=-17.9719921713269, pvalue=6.835532456916702e-54)
Ttest_indResult(statistic=-16.371375965888596, pvalue=1.0255175625864558e-46)
Ttest_indResult(statistic=1.677960615722372, pvalue=0.0940899398968142)
Ttest_indResult(statistic=-8.293603989407588, pvalue=1.5759322301054064e-15)
Ttest_indResult(statistic=20.414298768685285, pvalue=4.967619782338976e-64)
Ttest_indResult(statistic=13.448676866861932, pvalue=2.262878270897359e-34)
Ttest_indResu



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE